array/builder.go

```diff
@@ -2,53 +2,99 @@ package array
 
 import (
 	"bytes"
+	"sync/atomic"
+	"unsafe"
 
+	"github.com/apache/arrow/go/v7/arrow"
 	"github.com/apache/arrow/go/v7/arrow/array"
 	"github.com/apache/arrow/go/v7/arrow/bitutil"
 	"github.com/apache/arrow/go/v7/arrow/memory"
 )
 
 type StringBuilder struct {
-	mem      memory.Allocator
-	builder  *array.BinaryBuilder
-	constant bool
+	mem         memory.Allocator
+	len         int
+	cap         int
+	reserveData int
+	buffer      *memory.Buffer
+	builder     *array.BinaryBuilder
+	refCount    int64
 }
 
 func NewStringBuilder(mem memory.Allocator) *StringBuilder {
 	return &StringBuilder{
-		mem:      mem,
-		builder:  array.NewBinaryBuilder(mem, StringType),
-		constant: true,
+		mem:         mem,
+		len:         0,
+		cap:         0,
+		reserveData: 0,
+		buffer:      nil,
+		builder:     nil,
+		refCount:    1,
 	}
 }
 
 func (b *StringBuilder) Retain() {
-	b.builder.Retain()
+	atomic.AddInt64(&b.refCount, 1)
 }
 func (b *StringBuilder) Release() {
-	b.builder.Release()
+	if atomic.AddInt64(&b.refCount, -1) == 0 {
+		if b.buffer != nil {
+			b.buffer.Release()
+		}
+		if b.builder != nil {
+			b.builder.Release()
+		}
+	}
 }
 func (b *StringBuilder) Len() int {
-	return b.builder.Len()
+	if b.builder != nil {
+		return b.builder.Len()
+	}
+	return b.len
 }
 func (b *StringBuilder) Cap() int {
-	return b.builder.Cap()
+	if b.builder != nil {
+		return b.builder.Cap()
+	}
+	if b.cap > b.len {
+		return b.cap
+	}
+	return b.len
 }
 func (b *StringBuilder) NullN() int {
-	return b.builder.NullN()
+	if b.builder != nil {
+		return b.builder.NullN()
+	}
+	return 0
 }
 
 func (b *StringBuilder) AppendBytes(buf []byte) {
-	if b.builder.Len() > 0 {
-		b.constant = b.constant && bytes.Equal(buf, b.builder.Value(0))
+	if b.builder != nil {
+		b.builder.Append(buf)
+		return
+	}
+	if b.len == 0 {
+		b.buffer = memory.NewResizableBuffer(b.mem)
+		b.buffer.Resize(len(buf))
+		copy(b.buffer.Bytes(), buf)
+		b.len = 1
+		return
 	}
-	b.builder.Append(buf)
+	if bytes.Equal(b.buffer.Bytes(), buf) {
+		b.len++
+		return
+	}
+	b.makeBuilder(buf)
+
 }
 
 // Append appends a string to the array being built. The input string
 // will always be copied.
```

Comment:
Minor point: the second half of this comment was true prior to this change, but now isn't